### PR TITLE
Display backbones in UI

### PR DESCRIPTION
**Pull request by kdp-cloud**
This Pull Requests contains the implementation for visualizing the Backbones from the database in the User Interface.
It also contains some small changes (simplifications) in the model.

### CODE REVIEW

File: client/src/Main.elm [hunk 1]
```diff
@@ -410,13 +389,17 @@ view model =
         [ Element.layout
             [ Element.height Element.fill
             , inFront <| Notify.view CloseNotification model.notifications
+            , inFront <| navLinks
```
Comment (MaybeJustJames) [inline client/src/Main.elm:392]: Why is this necessary?

File: client/src/Main.elm [hunk 2]
```diff
@@ -1006,10 +894,10 @@ visualRepresentation model =
             sortByWith .bsa1_overhang ascending insertRecordList
 
         chartLabels =
-            List.reverse (model.selectedBackbone.name :: List.reverse (List.map .name sortedInsertRecordList))
+            List.reverse ((Maybe.withDefault "" <| Maybe.map .name model.selectedBackbone) :: List.reverse (List.map .name sortedInsertRecordList))
```
Comment (MaybeJustJames) [inline client/src/Main.elm:897]: Little nervous about 2x `List.reverse` here.

File: client/src/Main.elm [hunk 3]
```diff
@@ -1801,6 +1736,46 @@ level0Decoder =
...
+            case stringToBsmb1Overhang (String.trim bsmb1_str) of
+                Just bsmb1 ->
+                    Decode.succeed bsmb1
+
+                _ ->
+                    Decode.fail "Not a valid Gateway site"
+    in
+    Decode.succeed Backbone
+        |> JDP.required "name" Decode.string
+        |> JDP.required "id" (Decode.int |> Decode.map String.fromInt)
```
Comment (MaybeJustJames) [inline client/src/Main.elm:1762]: Huh? Surely this should be an `Int` then?

File: client/src/Main.elm [hunk 4]
```diff
@@ -1801,6 +1736,46 @@ level0Decoder =
...
+        |> JDP.optional "bsa1_overhang" (Decode.maybe (Decode.string |> Decode.andThen decodeOverhang)) Nothing
```
Comment (MaybeJustJames) [inline client/src/Main.elm:1763]: Is the `Decode.maybe` really required here?

File: client/src/Main.elm [hunk 5]
```diff
@@ -1898,12 +1876,52 @@ showOverhang bsa1_overhang =
             "F__G"
 
 
-stringToOverhang : String -> Maybe Overhang
-stringToOverhang strOverhang =
+showMaybeBsa1Overhang : Maybe Bsa1Overhang -> String
```
Comment (MaybeJustJames) [inline client/src/Main.elm:1879]: Why not `Bsa1Overhang -> Maybe String` as before?

**Review by MaybeJustJames — APPROVED** [review]
I approve in principle. Just want you to justify some things please.

File: client/src/Main.elm (hunk 1 above)
Comment (kdp-cloud) [reply client/src/Main.elm:392]: Because otherwise the navbar  isn't sticky anymore.

File: client/src/Main.elm (hunk 2 above)
Comment (kdp-cloud) [reply client/src/Main.elm:897]: This seemed like a good way to add the backbone to the beginning of the list.
Is there a better way?

File: client/src/Main.elm (hunk 5 above)
Comment (kdp-cloud) [reply client/src/Main.elm:1879]: For the level0 elements the  Bsa1Overhang is mandatory but for the backbones it isn't.
So I created a function for showing bsa1Overhangs in level0 elements:
```elm
showBsa1Overhang : Bsa1Overhang -> String
```
And a function for showing the bsa1Overhang in backbones:
```elm
showMaybeBsa1Overhang : Maybe Bsa1Overhang -> String
```

File: client/src/Main.elm (hunk 3 above)
Comment (kdp-cloud) [reply client/src/Main.elm:1762]: To be honest, I copied this from the level0Decoder we made together. So I figured it would have been a string here as well...

File: client/src/Main.elm [hunk 6]
```diff
@@ -242,20 +228,22 @@ init _ url key =
                 _ ->
                     getLoginUrls
     in
-    ( { page = Catalogue
+    ( { page = LoginPage
```
[the remark this answers is not in the document]
Comment (kdp-cloud) [reply client/src/Main.elm:231]: Thanks!

File: client/src/Main.elm (hunk 2 above)
Comment (MaybeJustJames) [reply client/src/Main.elm:897]: The final `reverse` puts it at the end though. Actually now I'm not sure why you're reversing at all

File: client/src/Main.elm (hunk 1 above)
Comment (MaybeJustJames) [reply client/src/Main.elm:392]: Ah ok :+1:

File: client/src/Main.elm (hunk 3 above)
Comment (kdp-cloud) [reply client/src/Main.elm]: I changed it to an ` Int` but I had to make some changes in the model and views.
I've pushed a new commit.

File: client/src/Main.elm [hunk 7]
```diff
@@ -38,15 +38,15 @@ import Url.Parser exposing ((</>), Parser, parse, query, s)
 import Url.Parser.Query exposing (map2, string)
 
 
-overhangList3 : List Overhang
+overhangList3 : List Bsa1Overhang
```
[the remark this answers is not in the document]
Comment (MaybeJustJames) [reply client/src/Main.elm:41]: FYI I've done this in the followup PR so feel free to ignore for now

File: client/src/Main.elm [hunk 8]
```diff
@@ -614,24 +595,16 @@ addBackboneView model =
...
             { onChange = ChangeBackboneNameToAdd
-            , text = model.nameBackboneToAdd
+            , text = Maybe.withDefault "" <| Maybe.map .name model.backboneToAdd
             , label = Input.labelLeft [] <| Element.text "Name:"
             , placeholder = Nothing
             }
         , Input.text []
             { onChange = ChangeBackboneMpgNumberToAdd
-            , text = model.mpgbNumberBackboneToAdd
+            , text = String.fromInt <| Maybe.withDefault 0 <| Maybe.map .mPGBNumber model.backboneToAdd
```
Comment (MaybeJustJames) [inline client/src/Main.elm:604]:
```suggestion
            , text =  Maybe.withDefault "" <| Maybe.map (String.fromInt << .mPGBNumber) model.backboneToAdd
```

File: client/src/Main.elm (hunk 4 above)
Comment (kdp-cloud) [reply client/src/Main.elm:1763]: I've changed it for the bsa1 overhang as well as for the bsmb1 overhang.

File: client/src/Main.elm (hunk 3 above)
Comment (kdp-cloud) [reply client/src/Main.elm]: I missed it. It should indeed be `mpg_number `and not `id`.

File: client/src/Main.elm (hunk 7 above)
[the remark this answers is not in the document]
Comment (kdp-cloud) [reply client/src/Main.elm]: It's changed.

File: client/src/Main.elm [hunk 9]
```diff
@@ -1357,11 +1246,15 @@ backboneTable model =
...
                     , { header = none
                       , width = fillPortion 1
-                      , view = .length >> String.fromInt >> Element.text >> el [ centerY ]
+                      , view = .sequence >> String.length >> String.fromInt >> Element.text >> el [ centerY ]
                       }
                     , { header = none
-                      , width = fillPortion 1
-                      , view = .level >> String.fromInt >> Element.text >> el [ centerY ]
+                      , width = fillPortion 3
+                      , view = .bsa1Overhang >> showMaybeBsa1Overhang >> Element.text >> el [ centerY ]
```
[the remark this answers is not in the document]
Comment (kdp-cloud) [reply client/src/Main.elm]: I've pushed the changes in a new commit.
I also changed it a bit so I can use the view for bsmb1 overhangs.

File: client/src/Main.elm [hunk 10]
```diff
@@ -1930,3 +1948,47 @@ stringToOverhang strOverhang =
...
+            "X__Y"
+
+        Just X__Z ->
+            "X__Z"
+
+        Just Y__Z ->
+            "Y__Z"
+
+        _ ->
+            "Not defined"
```
[the remark this answers is not in the document]
Comment (kdp-cloud) [reply client/src/Main.elm]: Resolved in previous remark.

File: client/src/Main.elm (hunk 8 above)
Comment (kdp-cloud) [reply client/src/Main.elm:577]: Done!
I've also done the same for the level 0 table.

File: client/src/Main.elm [hunk 11]
```diff
@@ -1357,11 +1246,15 @@ backboneTable model =
...
                       }
                     , { header = none
-                      , width = fillPortion 1
-                      , view = .level >> String.fromInt >> Element.text >> el [ centerY ]
+                      , width = fillPortion 3
+                      , view = .bsa1Overhang >> showMaybeBsa1Overhang >> Element.text >> el [ centerY ]
+                      }
+                    , { header = none
+                      , width = fillPortion 3
+                      , view = .bsmb1_overhang >> showMaybeBsmb1Overhang >> Element.text >> el [ centerY ]
```
[the remark this answers is not in the document]
Comment (kdp-cloud) [reply client/src/Main.elm]: Already fixed in previous remark

File: client/src/Main.elm [hunk 12]
```diff
@@ -1297,7 +1185,7 @@ insertTable model =
                       }
                     , { header = none
                       , width = fillPortion 1
-                      , view = .bsa1_overhang >> showOverhang >> Element.text >> el [ centerY ]
+                      , view = .bsa1_overhang >> showBsa1Overhang >> Element.text >> el [ centerY ]
```
[the remark this answers is not in the document]
Comment (kdp-cloud) [reply client/src/Main.elm]: Done!
